### PR TITLE
ENG-861: ADT backfill fast-follows

### PR DESCRIPTION
**Pull request by lucasdellabella**
### Dependencies

None

### Description

Adds a bunch of fast-follow items mentioned on [this PR](https://github.com/metriport/metriport/pull/4398) and [this PR](https://github.com/metriport/metriport/pull/4398).

This includes:
- correct script docstrings
- fix an edgecase in a pickMostSevereClass + add unit tests
- rename some stuff

### Testing

- Adds unit tests for `dangerouslyAssignMostSevereClass`

### Release Plan

- [ ] Merge this

## Summary by CodeRabbit

- **New Features**
  - Added a public ADT deduplication utility that performs datetime-aware encounter deduplication.
  - New script to locate ADT bundles lacking encounter resources.

- **Improvements**
  - More consistent status/class resolution across deduplication for multiple clinical resource types.

- **Documentation**
  - Updated usage instructions and commands for ADT bundle maintenance scripts.

- **Tests**
  - Expanded tests covering encounter class severity and edge cases.

- **Chores**
  - Introduced configurable dry-run and a confirmation pause for destructive maintenance scripts.

### CODE REVIEW

File: packages/utils/src/hl7v2-notifications/reprocess-adt-conversion-bundles/clean-duplicate-condition-ids.ts [hunk 1]
```diff
@@ -27,9 +27,10 @@ import { reprocessAdtConversionBundles } from "./common";
  */
 
 const prefixes: string[] = [];
+const dryRun = false;
```
Comment (leite08) [inline packages/utils/src/hl7v2-notifications/reprocess-adt-conversion-bundles/clean-duplicate-condition-ids.ts]: Can we add a note about it at the top of TSDoc to bring attention to it?

Or could require a cli param either true or false - to avoid running it with some default value by mistake.

Also, IMO all scripts that impact prod should have the validation/confirmation step - see `displayWarningAndConfirmation` on packages/utils.

It'd be good to standardize that ^ and having the datetime it started and how long it ran for. I can push a PR a wrapper for main files w/ those two later, but I'd have the confirmation as part of this PR.

File: packages/core/src/external/fhir/bundle/utils.ts [hunk 2]
```diff
@@ -37,13 +36,22 @@ export function mergeBundles({
...
+ * references will not be updated and will be left broken.
+ */
 export function dedupeAdtEncounters(existing: Bundle<Resource>): Bundle<Resource> {
   const [encounters, nonEncounters] = _(existing.entry)
     .map("resource")
     .partition(isEncounter)
     .value();
 
-  const { encountersMap: dedupedEncountersMap } = groupSameEncountersDateOnly(encounters);
+  const { encountersMap: dedupedEncountersMap } = groupSameEncountersDatetimeOnly(encounters);
```
Comment (leite08) [inline packages/core/src/external/fhir/bundle/utils.ts:56]: 😍

File: packages/core/src/fhir-deduplication/__tests__/shared.test.ts [hunk 3]
```diff
@@ -75,3 +76,101 @@ describe("pickMostSevereClass", () => {
     expect(result).toBeUndefined();
   });
 });
+
+describe("dangerouslyAssignMostSevereClass", () => {
```
Comment (leite08) [inline packages/core/src/fhir-deduplication/__tests__/shared.test.ts]: I suggest we have all unit tests inside file-wide "describe", so we can run all tests at once through the UI (clicking on "run" or "test").

We can have many levels, so we still group them by function.

See `error-mapping.test.ts`

File: packages/core/src/fhir-deduplication/shared.ts [hunk 4]
```diff
@@ -379,20 +379,29 @@ const classRanking = [
...
+ * Returns the rank of the class code in the classRanking array.
+ * The lower the rank, the more severe the class.
+ * @param c - The class coding to rank.
+ * @returns The rank of the class code in the classRanking array.
+ */
+function classRank(c: EncounterClassCoding | undefined) {
+  if (!c?.code) return Number.POSITIVE_INFINITY;
+  const idx = classRanking.indexOf(c.code.toUpperCase() as EncounterClassCode);
+  return idx === -1 ? Number.POSITIVE_INFINITY : idx;
+}
```
Comment (leite08) [inline packages/core/src/fhir-deduplication/shared.ts:392]: 💅

File: packages/utils/src/hl7v2-notifications/reprocess-adt-conversion-bundles/find-duplicate-resource-ids-within-a-bundle.ts [hunk 5]
```diff
@@ -27,35 +26,21 @@ import { reprocessAdtConversionBundles } from "./common";
...
-  await reprocessAdtConversionBundles(prefixes, cleanDuplicateConditionIds, true);
+  await reprocessAdtConversionBundles(prefixes, findAdtsWithoutEncounterResource, dryRun);
 }
 
-async function cleanDuplicateConditionIds(
-  bundle: FhirBundleSdk,
-  log: (message: string) => void
+async function findAdtsWithoutEncounterResource(
+  bundle: FhirBundleSdk
+  // log: (message: string) => void
```
Comment (leite08) [inline packages/utils/src/hl7v2-notifications/reprocess-adt-conversion-bundles/find-duplicate-resource-ids-within-a-bundle.ts]: left-over?

File: packages/utils/src/hl7v2-notifications/reprocess-adt-conversion-bundles/find-duplicate-resource-ids-within-a-bundle.ts [hunk 6]
```diff
@@ -27,35 +26,21 @@ import { reprocessAdtConversionBundles } from "./common";
...
-    .map(e => e.id)
-    .compact()
-    .value();
-
-  const encounterDuplicates = _.uniq(encounterIds).length !== encounterIds.length;
-
-  if (encounterDuplicates) {
-    log("Encounter duplicates found");
+  if (encounters.length === 0) {
+    process.stdout.write(".");
```
Comment (leite08) [inline packages/utils/src/hl7v2-notifications/reprocess-adt-conversion-bundles/find-duplicate-resource-ids-within-a-bundle.ts]: TIL

File: packages/utils/src/hl7v2-notifications/reprocess-adt-conversion-bundles/clean-duplicate-condition-ids.ts (hunk 1 above)
Comment (leite08) [reply packages/utils/src/hl7v2-notifications/reprocess-adt-conversion-bundles/clean-duplicate-condition-ids.ts]: here and on similar scripts

File: packages/core/src/external/fhir/bundle/utils.ts [hunk 7]
```diff
@@ -37,13 +36,22 @@ export function mergeBundles({
...
 }
 
+/**
+ * Additional deduplication logic for ADT encounters to more aggresively deduplicate than
+ * our core deduplication logic. This is only used for ADT encounters.
+ *
+ * ⚠️ NOTE: This function relies on the assumption that all resource ids in the bundle
+ * are stable, because we do not use the refReplacementMap from `groupSameEncountersDatetimeOnly`.
+ * If resource ids are not stable, and there are references pointing to an encounter, the
+ * references will not be updated and will be left broken.
```
Comment (leite08) [inline packages/core/src/external/fhir/bundle/utils.ts:46]: Aren't we going to address this behavior, so it accepts encounters w/ diff IDs? If later, can you add a ticket reference w/ TODO here please?

File: packages/utils/src/hl7v2-notifications/reprocess-adt-conversion-bundles/find-duplicate-resource-ids-within-a-bundle.ts [hunk 8]
```diff
@@ -4,14 +4,13 @@ dotenv.config();
...
 import { reprocessAdtConversionBundles } from "./common";
 
 /**
- * Finds duplicate resource IDs within ADT conversion bundles stored in S3.
+ * Finds ADT conversion bundles that don't contain any encounter resources.
  *
- * This script pulls down every conversion bundle from S3 and checks for duplicate resource IDs within each bundle.
- * It logs when duplicates are found but does not modify the bundles.
+ * This script pulls down every conversion bundle from S3 and checks if the bundle contains encounter resources.
+ * It prints a dot (.) for each bundle that doesn't have any encounters but does not modify the bundles.
```
Comment (leite08) [inline packages/utils/src/hl7v2-notifications/reprocess-adt-conversion-bundles/find-duplicate-resource-ids-within-a-bundle.ts]: Don't we need to know which ones are missing those? Could store the IDs on a file, in addition to logging the dot if that would be helpful. NABD, you've been running those and know what's more useful.

File: packages/core/src/external/fhir/bundle/utils.ts [hunk 9]
```diff
@@ -37,13 +36,22 @@ export function mergeBundles({
...
+ */
```
[the remark this answers is not in the document]
Comment (lucasdellabella) [reply packages/core/src/external/fhir/bundle/utils.ts]: excessive imo

File: packages/utils/src/hl7v2-notifications/reprocess-adt-conversion-bundles/clean-duplicate-condition-ids.ts (hunk 1 above)
Comment (lucasdellabella) [reply packages/utils/src/hl7v2-notifications/reprocess-adt-conversion-bundles/clean-duplicate-condition-ids.ts]: yep good idea, added it to the base script.

File: packages/core/src/fhir-deduplication/__tests__/shared.test.ts (hunk 3 above)
Comment (lucasdellabella) [reply packages/core/src/fhir-deduplication/__tests__/shared.test.ts]: addressed

File: packages/utils/src/hl7v2-notifications/reprocess-adt-conversion-bundles/find-duplicate-resource-ids-within-a-bundle.ts (hunk 8 above)
Comment (lucasdellabella) [reply packages/utils/src/hl7v2-notifications/reprocess-adt-conversion-bundles/find-duplicate-resource-ids-within-a-bundle.ts]: no, since imo this is a script thats just supposed to go through the array and print something if theres an issue. Makes it easy to see the quantity of issues, this was really just something i threw down for debugging purposes

Comment (lucasdellabella) [reply packages/utils/src/hl7v2-notifications/reprocess-adt-conversion-bundles/find-duplicate-resource-ids-within-a-bundle.ts]: I renamed it to make more sense. We can always use the provided logger to show the key of the file being processed, ill add that to the script tsdoc.

File: packages/utils/src/hl7v2-notifications/reprocess-adt-conversion-bundles/find-duplicate-resource-ids-within-a-bundle.ts (hunk 5 above)
Comment (lucasdellabella) [reply packages/utils/src/hl7v2-notifications/reprocess-adt-conversion-bundles/find-duplicate-resource-ids-within-a-bundle.ts]: yep

File: packages/core/src/external/fhir/bundle/utils.ts (hunk 7 above)
Comment (lucasdellabella) [reply packages/core/src/external/fhir/bundle/utils.ts:46]: yep, not addressing now since it doesn't materially affect anything and may not for a long time, or ever, and impact is low if it ever does i think.

File: packages/utils/src/hl7v2-notifications/reprocess-adt-conversion-bundles/common.ts [hunk 10]
```diff
@@ -8,6 +8,8 @@ import { Config } from "@metriport/core/util/config";
 import { out } from "@metriport/core/util/log";
 import { FhirBundleSdk } from "@metriport/fhir-sdk";
 import { parseFhirBundle } from "@metriport/shared/medical/fhir/bundle";
+import dayjs from "dayjs";
+import { sleep } from "@metriport/shared/common/sleep";
```
[the remark this answers is not in the document]
Comment (lucasdellabella) [reply packages/utils/src/hl7v2-notifications/reprocess-adt-conversion-bundles/common.ts]: addressed

File: packages/core/src/fhir-deduplication/__tests__/shared.test.ts [hunk 11]
```diff
@@ -1,77 +1,178 @@
...
+      const result = pickMostSevereClass(undefined, undefined);
+      expect(result).toBeUndefined();
+    });
+
+    it("returns undefined when both classes have no code", () => {
+      const class1 = {} as any;
+      const class2 = {} as any;
+      const result = pickMostSevereClass(class1, class2);
+      expect(result).toBeUndefined();
+    });
```
[the remark this answers is not in the document]
Comment (lucasdellabella) [reply packages/core/src/fhir-deduplication/__tests__/shared.test.ts:78]: addressed

**Review by leite08 — APPROVED** [review]
